feat(cohorts): Cache Cohort dependencies

dustinbyrne opened the pull request:
## Problem

This change introduces caching to forward and reverse dependency lookups for Cohorts. Identifying which Cohorts depend on any other given Cohort is an expensive calculation because it requires checking the properties of every other Cohort in the team.

This change is related to #36134. This PR focuses on cache warming and maintenance and a separate follow-up PR will implement usage of this cache for dependency recalculation.

## Changes
This PR introduces three main components:

1. Cohort dependency cache management module (`posthog/models/cohort/dependencies.py`)
2. Celery tasks for cache prewarming (`posthog/tasks/cohort_dependencies.py`)
    - Scheduled task that runs daily at 2 AM to warm caches for teams with many cohorts (>=50)
    - Prometheus metric to monitor active warming chains, for fine tuning if necessary
3. Renamed `get_dependent_cohorts` to `get_all_dependency_cohorts` for clarity

## How did you test this code?

Added automated tests including:
- Cache warming task behavior with different team sizes and batch configurations
- Pagination handling for large numbers of teams
- Error handling and retry logic
- Cache invalidation on cohort, team modifications
- Dependency extraction from cohort properties

haacked commented on posthog/models/cohort/dependencies.py:
```diff
@@ -0,0 +1,157 @@
+    """
+    dependencies = set()
+
+    for prop in cohort.properties.flat:
+        if prop.type == "cohort" and isinstance(prop.value, int):
+            dependencies.add(prop.value)
+
```
This doesn't handle the legacy `groups` format.

```suggestion
    for prop in cohort.properties.flat:
        if prop.type == "cohort" and isinstance(prop.value, int):
            dependencies.add(prop.value)

for group in cohort.groups:
    if not isinstance(group, dict):
        continue

    properties = group.get("properties", [])
    if isinstance(properties, list):
        for prop in properties:
            if isinstance(prop, dict) and prop.get("type") == "cohort":
                value = prop.get("value")
                if isinstance(value, int):
                    dependencies.add(value)
```

haacked commented on posthog/models/cohort/dependencies.py:
```diff
@@ -0,0 +1,157 @@
+    return dependencies
+
+
+def get_cohort_dependencies(cohort: Cohort) -> list[int]:
+    """
+    Get cohorts that this cohort depends on.
+    """
+    _warm_team_dependency_cache(cohort.team_id)
+    cache_key = f"cohort:deps:{cohort.id}"
+    return cache.get(cache_key, [])
```
This could lead to a race condition for a relatively expensive operation: `_warm_team_dependency_cache`. Consider refactoring this using `get_or_set`

```python
  def get_cohort_dependencies(cohort: Cohort) -> list[int]:
      """
      Get cohorts that this cohort depends on.
      """
      cache_key = f"cohort:deps:{cohort.id}"

      return cache.get_or_set(
          cache_key,
          lambda: list(extract_cohort_dependencies(cohort)),
          timeout=DEPENDENCY_CACHE_TIMEOUT
      )
```

And then actively maintain the cache using the signal handlers. I think this is simpler and doesn't need the team cache.

You could then add a daily scheduled task to `posthog/tasks/scheduled.py` (see other cache warmups in there). That would ensure the cache is nice and warm for end-users by the time they call `get_cohort_dependencies`.

haacked commented on posthog/models/cohort/dependencies.py:
```diff
@@ -0,0 +1,157 @@
+    # Just clear the team cache marker. The other keys will be overwritten the next time the cache is cleared.
+    cache.delete(f"team:deps_cached:{team_id}")
+
+
```
Make sure to add a signal handler for team deletion:

```python
@receiver(post_delete, sender="posthog.Team")
  def clear_team_cohort_dependency_cache(sender, instance, **kwargs):
      """Clear cohort dependency cache when team is deleted"""
      # Delete the cache entries for cohorts in this team.
  ```

haacked commented on posthog/models/cohort/dependencies.py:
```diff
@@ -0,0 +1,157 @@
+                refs.append(instance.id)
+                cache.set(refs_key, refs, timeout=DEPENDENCY_CACHE_TIMEOUT)
+
+        logger.debug(f"Updated dependency cache for cohort {instance.id}")
```
Make sure to run all this logic in `transaction.on_commit` like we do in `posthog/models/feature_flag/local_evaluation.py`.

One easy way to do this is to add an inner function that does the actual work and then call `transaction.on_commit(lambda: inner_function(...))`

haacked commented on posthog/models/cohort/dependencies.py:
```diff
@@ -0,0 +1,157 @@
+    """
+    Get cohorts that depend on the given cohort.
+    """
+    _warm_team_dependency_cache(cohort.team_id)
+    cache_key = f"cohort:refs:{cohort.id}"
+    return cache.get(cache_key, [])
```
See previous comment about `get_cohort_dependencies`. Same pattern applies here.

haacked commented on posthog/models/cohort/dependencies.py:
```diff
@@ -0,0 +1,157 @@
+    return cache.get(cache_key, [])
+
+
+def _warm_team_dependency_cache(team_id: int) -> None:
```
If you follow my earlier suggestions, this code should get a bit simpler. Something like:

```python
@receiver(post_save, sender="posthog.Cohort")
@receiver(post_delete, sender="posthog.Cohort")
def cohort_changed(sender, instance, **kwargs):
    """
    Clear and rebuild dependency caches when cohort changes.
    """
    def refresh_cache():
        try:
            # Clear this cohort's caches
            cache.delete(f"cohort:deps:{instance.id}")
            cache.delete(f"cohort:refs:{instance.id}")

            # Clear and rebuild reverse caches for all cohorts in the team
            team_cohorts = Cohort.objects.filter(team=instance.team, deleted=False)

            for cohort in team_cohorts:
                # Clear the reverse cache
                cache.delete(f"cohort:refs:{cohort.id}")
                # Immediately rebuild it
                get_cohort_dependents(cohort)  # This will repopulate via get_or_set

            # Skip warming for deleted cohorts
            if kwargs.get('signal').__name__ == 'post_delete':
                return

            # Rewarm this cohort's dependencies
            get_cohort_dependencies(instance)

        except Exception as e:
            logger.exception(f"Failed to refresh dependency cache for cohort {instance.id}",
error=e)

    transaction.on_commit(refresh_cache)
```

haacked reviewed (changes requested):
This is getting closer! We have a lot of existing caching in the app. One thing to look at is the HyperCache and how it's used. I don't think we need to use the HyperCache here. I think using the built-in `cache.get_or_set` should be sufficient. See my in-line comments to better understand what I mean.

dustinbyrne committed:
feat(cohorts): add dependency management and caching module

haacked reviewed (changes requested):
Almost there!

I think the big decision to make is how to handle cache warming. Here is a summary of my thoughts:

1. The complexity of cache warming for 7 teams is too much.
2. Spawning a celery task per 1,200+ teams is also too much.
3. A single celery task to warm all 1200 teams is probably fine.
4. Or, removing warming altogether is interesting, but poses a problem if folks don't visit or change cohorts very often, they might always get a cold cache. Unless you remove the cache TTL.
5. We should add some metrics. How long does the cache take to warm up? Cache hit/miss. Etc. The specific metrics depends on which approach you decide upon.

So I think the options are:

1. Lower the cache warming threshold to include more teams, but then the benefit is only for those teams.
2. Single celery task. No downsides other than the extra code and complexity of cache warming.
3. Remove cache warming and make TTL indefinite for cached cohort entries. This is simpler and ensures cache hits, but requires that our cache invalidation logic is complete. (If we do this, we should add a simple command `python manage.py clear_cohort_cache --team-id 123` to clear cache just in case)

What option do you want to go with?

dustinbyrne committed:
feat(cohorts): Add measurements for dependency cache hits/misses

dustinbyrne commented:
As discussed I've:
- Dropped the cache warming task
- Increased the TTL to one week
- Moved the recalculation work into an `on_commit` action to avoid any potential timing issues
- Added a test to ensure the cache is warmed before calculation occurs

dustinbyrne committed:
refactor: Rename `get_dependent_cohorts` to `get_all_cohort_dependencies`

Clarifies that the function returns all cohorts that the input cohort
depends on, not cohorts that depend on the input cohort.

dustinbyrne committed:
fix(cohorts): Recalculation occurs on transaction commit

This ensures that the Cohort is synchronized with the database before
recalculating and that other post_save signals have a chance to run.

haacked reviewed (approved):
Nice work! 🚢
